server/src/main/java/org/elasticsearch/indices/ShardLimitService.java

```diff
@@ -0,0 +1,135 @@
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(SETTING_CLUSTER_MAX_SHARDS_PER_NODE, this::setShardLimitPerNode);
+    }
+
+    // Protected for testing purposes
+    protected void setShardLimitPerNode(int newValue) {
+        this.shardLimitPerNode.set(newValue);
+    }
+
+    public int getShardLimitPerNode() {
+        return shardLimitPerNode.get();
+    }
+
+    /**
+     * Checks whether an index can be created without going over the cluster shard limit.
+     *
+     * @param settings       the settings of the index to be created
+     * @param state          the current cluster state
+     * @throws ValidationException if creating this index would put the cluster over the cluster shard limit
+     */
+    public void validateShardLimit(final Settings settings, final ClusterState state) {
+        final int numberOfShards = INDEX_NUMBER_OF_SHARDS_SETTING.get(settings);
+        final int numberOfReplicas = IndexMetadata.INDEX_NUMBER_OF_REPLICAS_SETTING.get(settings);
+        final int shardsToCreate = numberOfShards * (1 + numberOfReplicas);
+
+        final Optional<String> shardLimit = checkShardLimit(shardsToCreate, state);
+        if (shardLimit.isPresent()) {
+            final ValidationException e = new ValidationException();
+            e.addValidationError(shardLimit.get());
+            throw e;
+        }
+    }
+
+    /**
+     * Validates whether a list of indices can be opened without going over the cluster shard limit.  Only counts indices which are
+     * currently closed and will be opened, ignores indices which are already open.
+     *
+     * @param currentState The current cluster state.
+     * @param indices The indices which are to be opened.
+     * @throws ValidationException If this operation would take the cluster over the limit and enforcement is enabled.
+     */
+    public void validateShardLimit(ClusterState currentState, Index[] indices) {
+        int shardsToOpen = Arrays.stream(indices)
+            .filter(index -> currentState.metadata().index(index).getState().equals(IndexMetadata.State.CLOSE))
+            .mapToInt(index -> getTotalShardCount(currentState, index))
+            .sum();
+
+        Optional<String> error = checkShardLimit(shardsToOpen, currentState);
+        if (error.isPresent()) {
+            ValidationException ex = new ValidationException();
+            ex.addValidationError(error.get());
+            throw ex;
+        }
+    }
+
+    private static int getTotalShardCount(ClusterState state, Index index) {
+        IndexMetadata indexMetadata = state.metadata().index(index);
+        return indexMetadata.getNumberOfShards() * (1 + indexMetadata.getNumberOfReplicas());
+    }
+
+    /**
+     * Checks to see if an operation can be performed without taking the cluster over the cluster-wide shard limit. Adds a deprecation
+     * warning or returns an error message as appropriate
```

Comment:
I don't think any deprecation warning is added?

Reply:
This comment is referencing a deprecation warning that's present in the 7.x branch, not master. I'll remove the reference to the deprecation warning in master but we'll need to keep emitting the deprecation warning in the backport (which I believe there are tests for).